README.md
![logo](https://github.com/aloisdeniel/spaces/raw/master/logo.png)

# Spaces

<p>
  <a href="https://pub.dartlang.org/packages/spaces"><img src="https://img.shields.io/pub/v/spaces.svg"></a>
  <a href="https://www.buymeacoffee.com/aloisdeniel">
    <img src="https://img.shields.io/badge/$-donate-ff69b4.svg?maxAge=2592000&amp;style=flat">
  </a>
</p>

Defining global spacing constants for building consistent and responsive apps.

## Install

Add the dependency to your `pubspec.yaml` :

```yaml
dependencies:
    spaces: <version>
```

## Quick start

```dart
return MaterialApp(
    // Defining
    builder: (context, child) => Spacing(
        dataBuilder: (context) {
          final mediaQuery = MediaQuery.of(context);
          if (mediaQuery.size.width > 500) {
            return SpacingData.generate(30);
          }
          return SpacingData.generate(10);
        },
        child: child,
    ),
    home:Builder(builder: (context) {
        // Using
        final spacing = Spacing.of(context);
        return SpacedColumn.normal(
            padding: spacing.insets.all.semiBig,
            children: <Widget>[
                Text('hello'),
                Space.big(),
                Text('world'),
            ],
        );
      })
    ),
)
```

# Usage

## Configuring spacing 

A `Spacing` widget should declare a global spacing configuration for your application.

```dart
return MaterialApp(
    builder: (context, child) => Spacing.fixed(
        data: SpacingData.generate(10.0), // This generates a set of spacing constants for your apps, from a base space of `10.0`.
        child: child,
    ),
    /// ...
)
```

### Generate

By calling the `Spacing.generate` constructor, a set of constants are automatically generated from a single base value.

The generated values are :

* **extraSmall** : `value * 0.2`
* **small** : `value * 0.4`
* **semiSmall** : `value * 0.6`
* **normal** : `value`
* **semiBig** : `value * 1.5`
* **big** : `value * 2.5`
* **extraBig** : `value * 5.0`

```dart
SpacingData.generate(10.0)
```

### From spaces

If you want to define the various spacing values, you can use the 

```dart
SpacingData.fromSpaces(
    extraSmall: 1.0,
    small: 2.0,
    semiSmall: 4.0,
    normal: 8.0,
    semiBig: 12.0,
    big: 20.0,
    extraBig: 100.0,
)
```

## Responsive spacing

You can define the `dataBuilder` property of the `Spacing` constructor to generate a configuration from the given context.

```dart
Spacing(
    dataBuilder: (context) {
        final mediaQuery = MediaQuery.of(context);
        return SpacingData.generate(mediaQuery.size.width > 300.0 ? 20.0 : 10.0),
    },
    child: child,
)
```

## Using constants

To access various spacing constants, simply use the `Spacing.of(context)` getter, or the `context.spacing()` extension method.

```dart
final spacing = Spacing.of(context);
// spacing.spaces.semiBig
```

### Size variants

There are seven availables spacing constants : `extraSmall`, `small`, `semiSmall`, `normal`, `semiBig`, `big`, `extraBig`;

```dart
final spacing = Spacing.of(context);
return SizedBox(
    width: spacing.spaces.semiSmall,
);
```

### Insets helpers

You have `EdgeInsets` presets available from `<spacing>.insets.<selector>.<size>`.

The available selectors are : `all`, `horizontal`, `vertical`, `onlyLeft`, `onlyTop`, `onlyRight`, `onlyBottom`, `exceptLeft`, `exceptRight`, `exceptTop`, `exceptBottom`.

```dart
final spacing = Spacing.of(context);
return Padding(
    padding: spacing.insets.exceptLeft.semiBig,
    child: Text('Hello'),
);
```

## Widgets

### Space

The `Space` widget uses [gap](https://pub.dev/packages/gap) under the hood to define a space in a `Flex`, `Column`, `Row` or `Scrollable` (such as `ListView`).

```dart
return Column(
    children: [
        Text('hello'),
        const Space.semiSmall(),
        Text('world'),
    ],
);
```

### SpacedFlex / SpacedColumn / SpacedRow

Like `Flex` / `Column` / `Row`, but with additional `padding` and `spaceBetween` properties.

```dart
final spacing = Spacing.of(context);
return SpacedColumn.small(
    padding: spacing.insets.all.normal,
    children: [
        Text('hello'),
        Text('world'),
    ],
);
```

CHANGELOG.md
## [0.5.0] - 2021/03/04

* Migration to null-safety.

## [0.4.1] - 2020/08/20

* Updated generated values.
* Updated documentation.

## [0.4.0] - 2020/08/20

* Updated `SpacingData.fromSpaces` constructor.

## [0.3.1] - 2020/08/20

* Improved documentation with quickstart.

## [0.3.0] - 2020/08/20

* Updated generated amount.

## [0.2.2] - 2020/08/20

* Fixed sample in documentation

## [0.2.1] - 2020/08/20

* Fixed logo in documentation

## [0.2.0] - 2020/08/20

* Initial release


lib/src/space.dart
import 'package:flutter/widgets.dart';
import 'package:gap/gap.dart';

import 'spacing.dart';
import 'spacing_data.dart';

/// A widget that takes a fixed amount of space in the direction of its parent.
///
/// It only works in the following cases:
/// - It is a descendant of a [Row], [Column], or [Flex],
/// and the path from the [Gap] widget to its enclosing [Row], [Column], or
/// [Flex] must contain only [StatelessWidget]s or [StatefulWidget]s (not other
/// kinds of widgets, like [RenderObjectWidget]s).
/// - It is a descendant of a [Scrollable].
///
/// See also:
///
///  * [Gap], the base widget on which this widget is based upon
///  * [SliverSpace], the sliver version of this widget.
class Space extends StatelessWidget {
  /// The amount of space this widget takes in the direction of its parent.
  ///
  /// For example:
  /// - If the parent is a [Column] this is the height of this widget.
  /// - If the parent is a [Row] this is the width of this widget.
  ///
  /// Must not be null and must be positive.
  ///
  /// If null, then [size] should be defined.
  final double? mainAxisExtent;

  /// The amount of space this widget takes in the opposite direction of the
  /// parent.
  ///
  /// For example:
  /// - If the parent is a [Column] this is the width of this widget.
  /// - If the parent is a [Row] this is the height of this widget.
  ///
  /// Must be positive or null. If it's null (the default) the cross axis extent
  /// will be the same as the constraints of the parent in the opposite
  /// direction.
  final double? crossAxisExtent;

  /// The color used to fill the gap.
  final Color? color;

  /// The amount of space from [Spacing] this widget takes in the direction of its parent.
  ///
  /// For example:
  /// - If the parent is a [Column] this is the height of this widget.
  /// - If the parent is a [Row] this is the width of this widget.
  ///
  /// Must not be null and must be positive.
  ///
  /// If null, then [mainAxisExtent] should be defined.
  final SpaceSize? size;

  /// Creates a widget that takes a fixed [mainAxisExtent] of space in the
  /// direction of its parent.
  ///
  /// The [mainAxisExtent] must not be null and must be positive.
  /// The [crossAxisExtent] must be either null or positive.
  const Space({
    Key? key,
    required double mainAxisExtent,
    this.crossAxisExtent,
    this.color,
  })  : mainAxisExtent = mainAxisExtent,
        size = null,
        assert(mainAxisExtent > 0),
        super(key: key);

  /// Creates a widget that takes an `extraSmall` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.extraSmall({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.extraSmall,
        mainAxisExtent = null;

  /// Creates a widget that takes a `small` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.small({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.small,
        mainAxisExtent = null;

  /// Creates a widget that takes a `semiSmall` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.semiSmall({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.semiSmall,
        mainAxisExtent = null;

  /// Creates a widget that takes a `normal` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.normal({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.normal,
        mainAxisExtent = null;

  /// Creates a widget that takes a `semiBig` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.semiBig({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.semiBig,
        mainAxisExtent = null;

  /// Creates a widget that takes a `big` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.big({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.big,
        mainAxisExtent = null;

  /// Creates a widget that takes an `extraBig` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const Space.extraBig({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.extraBig,
        mainAxisExtent = null;

  @override
  Widget build(BuildContext context) {
    var mainAxisExtent = this.mainAxisExtent;
    if (mainAxisExtent == null) {
      var size = this.size ?? SpaceSize.normal;
      mainAxisExtent = size.toPoints(context);
    }
    return Gap(
      mainAxisExtent,
      crossAxisExtent: crossAxisExtent,
      color: color,
    );
  }
}

/// A sliver that takes a fixed amount of space.
///
/// See also:
///
///  * [Space], the render box version of this widget.
class SliverSpace extends StatelessWidget {
  /// The amount of space this widget takes in the direction of its parent.
  ///
  /// For example:
  /// - If the parent is a [Column] this is the height of this widget.
  /// - If the parent is a [Row] this is the width of this widget.
  ///
  /// Must not be null and must be positive.
  ///
  /// If null, then [size] should be defined.
  final double? mainAxisExtent;

  /// The amount of space this widget takes in the opposite direction of the
  /// parent.
  ///
  /// Must be positive or null. If it's null (the default) the cross axis extent
  /// will be the same as the constraints of the parent in the opposite
  /// direction.
  final double? crossAxisExtent;

  /// The color used to fill the gap.
  final Color? color;

  /// The amount of space from [Spacing] this widget takes in the direction of its parent.
  ///
  /// Must not be null and must be positive.
  ///
  /// If null, then [mainAxisExtent] should be defined.
  final SpaceSize? size;

  /// Creates a widget that takes a fixed [mainAxisExtent] of space in the
  /// direction of its parent.
  ///
  /// The [mainAxisExtent] must not be null and must be positive.
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace({
    Key? key,
    required double mainAxisExtent,
    this.crossAxisExtent,
    this.color,
  })  : mainAxisExtent = mainAxisExtent,
        size = null,
        assert(mainAxisExtent > 0),
        super(key: key);

  /// Creates a widget that takes an `extraSmall` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.extraSmall({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.extraSmall,
        mainAxisExtent = null;

  /// Creates a widget that takes a `small` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.small({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.small,
        mainAxisExtent = null;

  /// Creates a widget that takes a `semiSmall` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.semiSmall({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.semiSmall,
        mainAxisExtent = null;

  /// Creates a widget that takes a `normal` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.normal({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.normal,
        mainAxisExtent = null;

  /// Creates a widget that takes a `semiBig` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.semiBig({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.semiBig,
        mainAxisExtent = null;

  /// Creates a widget that takes a `big` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.big({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.big,
        mainAxisExtent = null;

  /// Creates a widget that takes an `extraBig` amount of space in the
  /// direction of its parent.
  ///
  /// The spacing value is defined in the global [Spacing] definition.
  ///
  /// The [crossAxisExtent] must be either null or positive.
  const SliverSpace.extraBig({
    this.crossAxisExtent,
    this.color,
  })  : size = SpaceSize.extraBig,
        mainAxisExtent = null;

  @override
  Widget build(BuildContext context) {
    var mainAxisExtent = this.mainAxisExtent;
    if (mainAxisExtent == null) {
      var size = this.size ?? SpaceSize.normal;
      mainAxisExtent = size.toPoints(context);
    }
    return SliverGap(mainAxisExtent);
  }
}


lib/src/spacing.dart
import 'package:flutter/widgets.dart';
import 'package:spaces/src/spacing_data.dart';

typedef DataBuilder<T> = T Function(BuildContext context);

/// Applies a global spacing configuration to descendant widgets.
///
/// See also:
///
///  * [SpacingData], which describes the actual spacing configuration
///  * [Space], a gap widget that uses the spacing configuration
///  * [SpacedFlex], a flex widget that uses the spacing configuration
class Spacing extends StatelessWidget {
  final DataBuilder<SpacingData> dataBuilder;
  final Widget child;

  /// Applies the spacing resulting from the given [dataBuilder] to [child].
  ///
  /// The [dataBuilder] and [child] arguments must not be null.
  const Spacing({
    Key? key,
    required this.child,
    required this.dataBuilder,
  }) : super(key: key);

  /// Applies the given [data] to [child].
  ///
  /// The [data] and [child] arguments must not be null.
  Spacing.fixed({
    Key? key,
    required this.child,
    required SpacingData data,
  })   : assert(data != null),
        assert(child != null),
        dataBuilder = ((BuildContext context) => data),
        super(key: key);

  /// The data from the closest [Spacing] instance that encloses the given
  /// context.
  ///
  /// Typical usage is as follows:
  ///
  /// ```dart
  /// @override
  /// Widget build(BuildContext context) {
  ///   return SizedBox(
  ///     height: Spacing.of(context).spaces.small,
  ///   );
  /// }
  /// ```
  ///
  /// When the [Spacing] is actually created in the same `build` function
  /// (possibly indirectly, e.g. as part of a [MaterialApp]), the `context`
  /// argument to the `build` function can't be used to find the [Spacing] (since
  /// it's "above" the widget being returned). In such cases, the following
  /// technique with a [Builder] can be used to provide a new scope with a
  /// [BuildContext] that is "under" the [Spacing]:
  ///
  /// ```dart
  /// @override
  /// Widget build(BuildContext context) {
  ///   return Spacing(
  ///     data: SpacingData.fixed(10.0),
  ///     body: Builder(
  ///       // Create an inner BuildContext so that we can refer to
  ///       // the Spacing with Spacing.of().
  ///       builder: (BuildContext context) {
  ///         return Center(
  ///           child:SizedBox(
  ///             height: Spacing.of(context).spaces.small,
  ///           ),
  ///         );
  ///       },
  ///     ),
  ///   );
  /// }
  /// ```
  static SpacingData of(BuildContext context) {
    final provider =
        context.dependOnInheritedWidgetOfExactType<_SpacingDataProvider>();
    assert(
      provider != null,
      'No Spacing in widget tree. Make sure that you declared a Spacing.',
    );
    return provider!.data;
  }

  @override
  Widget build(BuildContext context) {
    return _SpacingDataProvider(
      data: dataBuilder(context),
      child: child,
    );
  }
}

extension SpacingExtension on BuildContext {
  SpacingData spacing() => Spacing.of(this);
}

extension SpaceSizeExtension on SpaceSize {
  double toPoints(BuildContext context) {
    final spacing = context.spacing();
    switch (this) {
      case SpaceSize.extraSmall:
        return spacing.spaces.extraSmall;
      case SpaceSize.small:
        return spacing.spaces.small;
      case SpaceSize.semiSmall:
        return spacing.spaces.semiSmall;
      case SpaceSize.semiBig:
        return spacing.spaces.semiBig;
      case SpaceSize.big:
        return spacing.spaces.big;
      case SpaceSize.extraBig:
        return spacing.spaces.extraBig;
      default:
        return spacing.spaces.normal;
    }
  }
}

class _SpacingDataProvider extends InheritedWidget {
  final SpacingData data;
  const _SpacingDataProvider({
    Key? key,
    required Widget child,
    required this.data,
  }) : super(
          key: key,
          child: child,
        );

  @override
  bool updateShouldNotify(_SpacingDataProvider oldWidget) {
    return data != oldWidget.data;
  }
}


lib/src/spacing_data.dart
import 'package:flutter/widgets.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'spacing_data.freezed.dart';

/// Defines the configuration of the overall visual [Spacing] for a
/// widget subtree within the app.
@freezed
abstract class SpacingData with _$SpacingData {
  /// Generates all [spaces] and [insets] from a given
  /// [normal] base space value.
  ///
  /// Several types of spaces are available :
  ///
  /// * [spaces] : `double` values
  /// * [insets] : `EdgeInsets` values
  ///
  /// The values follows this calculation :
  ///
  //  * extraSmall: `normal * 0.2`
  //  * small: `normal * 0.4`
  //  * semiSmall: `normal * 0.8`
  //  * semiBig: `normal * 1.5`
  //  * big: `normal * 2.5`
  //  * extraBig: `normal * 5.0`
  factory SpacingData.generate(double normal) {
    return SpacingData.fromSpaces(
      extraSmall: normal * 0.2,
      small: normal * 0.4,
      semiSmall: normal * 0.6,
      normal: normal,
      semiBig: normal * 1.5,
      big: normal * 2.5,
      extraBig: normal * 5.0,
    );
  }

  /// Generates all [insets] from the given [spaces].
  factory SpacingData.fromSpaces({
    required double extraSmall,
    required double small,
    required double semiSmall,
    required double normal,
    required double semiBig,
    required double big,
    required double extraBig,
  }) {
    final spaces = SizeSet(
      extraSmall: extraSmall,
      small: small,
      semiSmall: semiSmall,
      normal: normal,
      semiBig: semiBig,
      big: big,
      extraBig: extraBig,
    );
    return SpacingData(
      spaces: spaces,
      insets: SpacingInsetsData.fromSpaces(spaces),
    );
  }

  /// A spacing configuration from the given [spaces] and [insets].
  const factory SpacingData({
    required SpacingInsetsData insets,
    required SizeSet<double> spaces,
  }) = _SpacingData;
}

/// Defines the edge insets configuration.
///
/// A set of presets are available :
///
/// * [all] equivalent of `EdgeInsets.all(<size>)`
/// * [vertical] equivalent of `EdgeInsets.symmetric(vertical: <size>)`
/// * [horizontal] equivalent of `EdgeInsets.symmetric(horizontal: <size>)`
/// * [onlyRight] equivalent of `EdgeInsets.only(right: <size>)`
/// * [onlyLeft] equivalent of `EdgeInsets.only(left: <size>)`
/// * [onlyTop] equivalent of `EdgeInsets.only(top: <size>)`
/// * [onlyBottom] equivalent of `EdgeInsets.only(bottom: <size>)`
/// * [exceptRight] equivalent of `EdgeInsets.fromLTRB(<size>,<size>,0,<size>)`
/// * [exceptLeft] equivalent of `EdgeInsets.fromLTRB(0,<size>,<size>,<size>)`
/// * [exceptTop] equivalent of `EdgeInsets.fromLTRB(<size>,0,<size>,<size>)`
/// * [exceptBottom] equivalent of `EdgeInsets.fromLTRB(<size>,<size>,<size>,0)`
@freezed
abstract class SpacingInsetsData with _$SpacingInsetsData {
  factory SpacingInsetsData.fromSpaces(SizeSet<double> spaces) {
    return SpacingInsetsData(
      all: SizeSet(
        extraSmall: EdgeInsets.all(spaces.extraSmall),
        small: EdgeInsets.all(spaces.small),
        semiSmall: EdgeInsets.all(spaces.semiSmall),
        normal: EdgeInsets.all(spaces.normal),
        semiBig: EdgeInsets.all(spaces.semiBig),
        big: EdgeInsets.all(spaces.big),
        extraBig: EdgeInsets.all(spaces.big),
      ),
      vertical: SizeSet(
        extraSmall: EdgeInsets.symmetric(vertical: spaces.extraSmall),
        small: EdgeInsets.symmetric(vertical: spaces.small),
        semiSmall: EdgeInsets.symmetric(vertical: spaces.semiSmall),
        normal: EdgeInsets.symmetric(vertical: spaces.normal),
        semiBig: EdgeInsets.symmetric(vertical: spaces.semiBig),
        big: EdgeInsets.symmetric(vertical: spaces.big),
        extraBig: EdgeInsets.symmetric(vertical: spaces.big),
      ),
      horizontal: SizeSet(
        extraSmall: EdgeInsets.symmetric(horizontal: spaces.extraSmall),
        small: EdgeInsets.symmetric(horizontal: spaces.small),
        semiSmall: EdgeInsets.symmetric(horizontal: spaces.semiSmall),
        normal: EdgeInsets.symmetric(horizontal: spaces.normal),
        semiBig: EdgeInsets.symmetric(horizontal: spaces.semiBig),
        big: EdgeInsets.symmetric(horizontal: spaces.big),
        extraBig: EdgeInsets.symmetric(horizontal: spaces.big),
      ),
      onlyRight: SizeSet(
        extraSmall: EdgeInsets.only(right: spaces.extraSmall),
        small: EdgeInsets.only(right: spaces.small),
        semiSmall: EdgeInsets.only(right: spaces.semiSmall),
        normal: EdgeInsets.only(right: spaces.normal),
        semiBig: EdgeInsets.only(right: spaces.semiBig),
        big: EdgeInsets.only(right: spaces.big),
        extraBig: EdgeInsets.only(right: spaces.big),
      ),
      onlyLeft: SizeSet(
        extraSmall: EdgeInsets.only(left: spaces.extraSmall),
        small: EdgeInsets.only(left: spaces.small),
        semiSmall: EdgeInsets.only(left: spaces.semiSmall),
        normal: EdgeInsets.only(left: spaces.normal),
        semiBig: EdgeInsets.only(left: spaces.semiBig),
        big: EdgeInsets.only(left: spaces.big),
        extraBig: EdgeInsets.only(left: spaces.big),
      ),
      onlyBottom: SizeSet(
        extraSmall: EdgeInsets.only(bottom: spaces.extraSmall),
        small: EdgeInsets.only(bottom: spaces.small),
        semiSmall: EdgeInsets.only(bottom: spaces.semiSmall),
        normal: EdgeInsets.only(bottom: spaces.normal),
        semiBig: EdgeInsets.only(bottom: spaces.semiBig),
        big: EdgeInsets.only(bottom: spaces.big),
        extraBig: EdgeInsets.only(bottom: spaces.big),
      ),
      onlyTop: SizeSet(
        extraSmall: EdgeInsets.only(top: spaces.extraSmall),
        small: EdgeInsets.only(top: spaces.small),
        semiSmall: EdgeInsets.only(top: spaces.semiSmall),
        normal: EdgeInsets.only(top: spaces.normal),
        semiBig: EdgeInsets.only(top: spaces.semiBig),
        big: EdgeInsets.only(top: spaces.big),
        extraBig: EdgeInsets.only(top: spaces.big),
      ),
      exceptBottom: SizeSet(
        extraSmall: EdgeInsets.fromLTRB(
          spaces.extraSmall,
          spaces.extraSmall,
          spaces.extraSmall,
          0,
        ),
        small: EdgeInsets.fromLTRB(
          spaces.small,
          spaces.small,
          spaces.small,
          0,
        ),
        semiSmall: EdgeInsets.fromLTRB(
          spaces.semiSmall,
          spaces.semiSmall,
          spaces.semiSmall,
          0,
        ),
        normal: EdgeInsets.fromLTRB(
          spaces.normal,
          spaces.normal,
          spaces.normal,
          0,
        ),
        semiBig: EdgeInsets.fromLTRB(
          spaces.semiBig,
          spaces.semiBig,
          spaces.semiBig,
          0,
        ),
        big: EdgeInsets.fromLTRB(
          spaces.big,
          spaces.big,
          spaces.big,
          0,
        ),
        extraBig: EdgeInsets.fromLTRB(
          spaces.extraBig,
          spaces.extraBig,
          spaces.extraBig,
          0,
        ),
      ),
      exceptLeft: SizeSet(
        extraSmall: EdgeInsets.fromLTRB(
          0,
          spaces.extraSmall,
          spaces.extraSmall,
          spaces.extraSmall,
        ),
        small: EdgeInsets.fromLTRB(
          0,
          spaces.small,
          spaces.small,
          spaces.small,
        ),
        semiSmall: EdgeInsets.fromLTRB(
          0,
          spaces.semiSmall,
          spaces.semiSmall,
          spaces.semiSmall,
        ),
        normal: EdgeInsets.fromLTRB(
          0,
          spaces.normal,
          spaces.normal,
          spaces.normal,
        ),
        semiBig: EdgeInsets.fromLTRB(
          0,
          spaces.semiBig,
          spaces.semiBig,
          spaces.semiBig,
        ),
        big: EdgeInsets.fromLTRB(
          0,
          spaces.big,
          spaces.big,
          spaces.big,
        ),
        extraBig: EdgeInsets.fromLTRB(
          0,
          spaces.extraBig,
          spaces.extraBig,
          spaces.extraBig,
        ),
      ),
      exceptRight: SizeSet(
        extraSmall: EdgeInsets.fromLTRB(
          spaces.extraSmall,
          spaces.extraSmall,
          0,
          spaces.extraSmall,
        ),
        small: EdgeInsets.fromLTRB(
          spaces.small,
          spaces.small,
          0,
          spaces.small,
        ),
        semiSmall: EdgeInsets.fromLTRB(
          spaces.semiSmall,
          spaces.semiSmall,
          0,
          spaces.semiSmall,
        ),
        normal: EdgeInsets.fromLTRB(
          spaces.normal,
          spaces.normal,
          0,
          spaces.normal,
        ),
        semiBig: EdgeInsets.fromLTRB(
          spaces.semiBig,
          spaces.semiBig,
          0,
          spaces.semiBig,
        ),
        big: EdgeInsets.fromLTRB(
          spaces.big,
          spaces.big,
          0,
          spaces.big,
        ),
        extraBig: EdgeInsets.fromLTRB(
          spaces.extraBig,
          spaces.extraBig,
          0,
          spaces.extraBig,
        ),
      ),
      exceptTop: SizeSet(
        extraSmall: EdgeInsets.fromLTRB(
          spaces.extraSmall,
          0,
          spaces.extraSmall,
          spaces.extraSmall,
        ),
        small: EdgeInsets.fromLTRB(
          spaces.small,
          0,
          spaces.small,
          spaces.small,
        ),
        semiSmall: EdgeInsets.fromLTRB(
          spaces.semiSmall,
          0,
          spaces.semiSmall,
          spaces.semiSmall,
        ),
        normal: EdgeInsets.fromLTRB(
          spaces.normal,
          0,
          spaces.normal,
          spaces.normal,
        ),
        semiBig: EdgeInsets.fromLTRB(
          spaces.semiBig,
          0,
          spaces.semiBig,
          spaces.semiBig,
        ),
        big: EdgeInsets.fromLTRB(
          spaces.big,
          0,
          spaces.big,
          spaces.big,
        ),
        extraBig: EdgeInsets.fromLTRB(
          spaces.extraBig,
          0,
          spaces.extraBig,
          spaces.extraBig,
        ),
      ),
    );
  }

  const factory SpacingInsetsData({
    required SizeSet<EdgeInsets> all,
    required SizeSet<EdgeInsets> horizontal,
    required SizeSet<EdgeInsets> vertical,
    required SizeSet<EdgeInsets> onlyRight,
    required SizeSet<EdgeInsets> onlyTop,
    required SizeSet<EdgeInsets> onlyBottom,
    required SizeSet<EdgeInsets> onlyLeft,
    required SizeSet<EdgeInsets> exceptLeft,
    required SizeSet<EdgeInsets> exceptRight,
    required SizeSet<EdgeInsets> exceptTop,
    required SizeSet<EdgeInsets> exceptBottom,
  }) = _SpacingInsetsData;
}

/// A set of size variants, from [extraSmall] to [extraBig]
@freezed
abstract class SizeSet<T> with _$SizeSet<T> {
  /// The user is currently updating the login form.
  const factory SizeSet({
    required T extraSmall,
    required T small,
    required T semiSmall,
    required T normal,
    required T semiBig,
    required T big,
    required T extraBig,
  }) = _SizeSet;
}

enum SpaceSize {
  extraSmall,
  small,
  semiSmall,
  normal,
  semiBig,
  big,
  extraBig,
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:spaces/spaces.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Spacing Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      builder: (context, child) => Spacing(
        dataBuilder: (context) {
          final mediaQuery = MediaQuery.of(context);
          if (mediaQuery.size.width > 500) {
            return SpacingData.generate(30);
          }
          return SpacingData.generate(10);
        },
        child: child ?? SizedBox(),
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 3,
      child: Scaffold(
        appBar: AppBar(
          title: Text('Spacing'),
          bottom: TabBar(
            tabs: [
              Tab(
                text: 'Values',
              ),
              Tab(
                text: 'Space',
              ),
              Tab(
                text: 'SpacedFlex',
              ),
            ],
          ),
        ),
        body: TabBarView(
          children: [
            const ValuesTab(),
            const SpaceTab(),
            const SpacedFlexTab(),
          ],
        ),
      ),
    );
  }
}

class ValuesTab extends StatelessWidget {
  const ValuesTab();

  Widget _box(BuildContext context, String title, double height) {
    final spacing = Spacing.of(context);
    return Padding(
      padding: spacing.insets.all.normal,
      child: Column(
        children: [
          Text(
            title,
            textAlign: TextAlign.center,
          ),
          Container(
            color: Colors.red,
            width: spacing.spaces.big,
            height: height,
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final spacing = Spacing.of(context);
    return Container(
      color: Colors.red.withOpacity(0.2),
      margin: spacing.insets.all.big,
      padding: spacing.insets.all.normal,
      child: ListView(
        children: <Widget>[
          _box(context, 'extraSmall', spacing.spaces.extraSmall),
          _box(context, 'small', spacing.spaces.small),
          _box(context, 'semiSmall', spacing.spaces.semiSmall),
          _box(context, 'normal', spacing.spaces.normal),
          _box(context, 'semiBig', spacing.spaces.semiBig),
          _box(context, 'big', spacing.spaces.big),
          _box(context, 'extraBig', spacing.spaces.extraBig),
        ],
      ),
    );
  }
}

class SpaceTab extends StatelessWidget {
  const SpaceTab();

  Widget _box(BuildContext context) {
    return Container(
      color: Colors.blue,
      width: 20.0,
      height: 20.0,
    );
  }

  @override
  Widget build(BuildContext context) {
    final spacing = Spacing.of(context);
    return Container(
      color: Colors.blue.withOpacity(0.2),
      margin: spacing.insets.all.big,
      padding: spacing.insets.all.normal,
      child: ListView(
        children: <Widget>[
          _box(context),
          Space.extraSmall(),
          _box(context),
          Space.small(),
          _box(context),
          Space.semiSmall(),
          _box(context),
          Space.normal(),
          _box(context),
          Space.semiBig(),
          _box(context),
          Space.big(),
          _box(context),
          Space.extraBig(),
          _box(context),
        ],
      ),
    );
  }
}

class SpacedFlexTab extends StatelessWidget {
  const SpacedFlexTab();

  Widget _box(BuildContext context) {
    return Container(
      color: Colors.green,
      width: 20.0,
      height: 20.0,
    );
  }

  @override
  Widget build(BuildContext context) {
    final spacing = Spacing.of(context);
    return Container(
      color: Colors.green.withOpacity(0.2),
      child: ListView(
        children: [
          SpacedColumn.normal(
            padding: spacing.insets.all.normal,
            children: <Widget>[
              _box(context),
              _box(context),
              _box(context),
              _box(context),
            ],
          ),
          Center(
            child: SpacedRow.normal(
              padding: spacing.insets.all.normal,
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                _box(context),
                _box(context),
                _box(context),
                _box(context),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
